File: src/main/java/com/sourcegraph/cody/autocomplete/CodyEditorFactoryListener.java
package com.sourcegraph.cody.autocomplete;

import static com.sourcegraph.utils.CodyEditorUtil.VIM_EXIT_INSERT_MODE_ACTION;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.command.CommandProcessor;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.SelectionModel;
import com.intellij.openapi.editor.VisualPosition;
import com.intellij.openapi.editor.event.*;
import com.intellij.openapi.editor.ex.util.EditorUtil;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.fileEditor.FileEditor;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.fileEditor.TextEditor;
import com.intellij.openapi.fileEditor.impl.FileEditorManagerImpl;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Disposer;
import com.intellij.openapi.vfs.VirtualFile;
import com.sourcegraph.cody.agent.CodyAgent;
import com.sourcegraph.cody.agent.CodyAgentClient;
import com.sourcegraph.cody.agent.protocol.Position;
import com.sourcegraph.cody.agent.protocol.Range;
import com.sourcegraph.cody.agent.protocol.TextDocument;
import com.sourcegraph.cody.vscode.InlineAutocompleteTriggerKind;
import com.sourcegraph.cody.vscode.InlineCompletionTriggerKind;
import com.sourcegraph.config.ConfigUtil;
import com.sourcegraph.utils.CodyEditorUtil;
import java.util.List;
import java.util.Objects;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Determines when to trigger completions and when to clear completions.
 *
 * <p>IntelliJ doesn't have a built-in API to register "inline completion providers" similar to VS
 * Code. Instead, we manually listen to editor events like the caret position, selection changes,
 * and document edits.
 */
public class CodyEditorFactoryListener implements EditorFactoryListener {
  CodySelectionListener selectionListener = new CodySelectionListener();
  CaretListener caretListener = new CodyCaretListener();

  @Override
  public void editorCreated(@NotNull EditorFactoryEvent event) {
    if (!ConfigUtil.isCodyEnabled()) {
      return;
    }
    Editor editor = event.getEditor();
    informAgentAboutEditorChange(editor, false);
    Project project = editor.getProject();
    if (project == null || project.isDisposed()) {
      return;
    }
    Disposable disposable = Disposer.newDisposable("CodyEditorFactoryListener");
    EditorUtil.disposeWithEditor(editor, disposable);
    editor.getCaretModel().addCaretListener(this.caretListener, disposable);
    editor.getSelectionModel().addSelectionListener(this.selectionListener, disposable);
    editor.getDocument().addDocumentListener(new CodyDocumentListener(editor), disposable);
  }

  private static class CodyCaretListener implements CaretListener {

    @Override
    public void caretPositionChanged(@NotNull CaretEvent e) {
      if (!ConfigUtil.isCodyEnabled()) {
        return;
      }
      String commandName = CommandProcessor.getInstance().getCurrentCommandName();
      if (Objects.equals(commandName, VIM_EXIT_INSERT_MODE_ACTION)) {
        return;
      }
      informAgentAboutEditorChange(e.getEditor(), true);
      CodyAutocompleteManager suggestions = CodyAutocompleteManager.getInstance();
      Editor editor = e.getEditor();
      if (CodyEditorUtil.isEditorValidForAutocomplete(editor)
          && CodyEditorFactoryListener.isSelectedEditor(editor)) {
        suggestions.clearAutocompleteSuggestions(e.getEditor());
        if (CodyEditorUtil.isImplicitAutocompleteEnabledForEditor(editor))
          suggestions.triggerAutocomplete(
              e.getEditor(),
              e.getEditor().getCaretModel().getOffset(),
              InlineCompletionTriggerKind.AUTOMATIC);
      }
    }
  }

  private static class CodySelectionListener implements SelectionListener {
    @Override
    public void selectionChanged(@NotNull SelectionEvent e) {
      if (!ConfigUtil.isCodyEnabled()) {
        return;
      }
      Editor editor = e.getEditor();
      informAgentAboutEditorChange(editor, true);
      if (CodyEditorUtil.isEditorValidForAutocomplete(editor)
          && ConfigUtil.isCodyEnabled()
          && CodyEditorFactoryListener.isSelectedEditor(editor))
        CodyAutocompleteManager.getInstance().clearAutocompleteSuggestions(editor);
    }
  }

  private static class CodyDocumentListener implements BulkAwareDocumentListener {
    private final Editor editor;

    public CodyDocumentListener(@NotNull Editor editor) {
      this.editor = editor;
    }

    public void documentChangedNonBulk(@NotNull DocumentEvent event) {
      if (!CodyEditorFactoryListener.isSelectedEditor(this.editor)) {
        return;
      }
      CodyAutocompleteManager completions = CodyAutocompleteManager.getInstance();
      completions.clearAutocompleteSuggestions(this.editor);
      if (CodyEditorUtil.isImplicitAutocompleteEnabledForEditor(this.editor)
          && CodyEditorUtil.isEditorValidForAutocomplete(this.editor)
          && !CommandProcessor.getInstance().isUndoTransparentActionInProgress()) {
        informAgentAboutEditorChange(this.editor, true);
        int changeOffset = event.getOffset() + event.getNewLength();
        if (this.editor.getCaretModel().getOffset() == changeOffset) {
          InlineAutocompleteTriggerKind requestType =
              event.getOldLength() != event.getNewLength()
                  ? InlineAutocompleteTriggerKind.Invoke
                  : InlineAutocompleteTriggerKind.Automatic;
          completions.triggerAutocomplete(
              this.editor, changeOffset, InlineCompletionTriggerKind.AUTOMATIC);
        }
      }
    }
  }

  /**
   * Returns true if this editor is currently open and focused by the user. Returns true if this
   * editor is in a separate tab or not focused/selected by the user.
   */
  private static boolean isSelectedEditor(Editor editor) {
    if (editor == null) {
      return false;
    }
    Project project = editor.getProject();
    if (project == null || project.isDisposed()) {
      return false;
    }
    FileEditorManager editorManager = FileEditorManager.getInstance(project);
    if (editorManager == null) {
      return false;
    }
    if (editorManager instanceof FileEditorManagerImpl) {
      Editor current = ((FileEditorManagerImpl) editorManager).getSelectedTextEditor(true);
      return current != null && current.equals(editor);
    }
    FileEditor current = editorManager.getSelectedEditor();
    return current instanceof TextEditor && editor.equals(((TextEditor) current).getEditor());
  }

  @Nullable
  private static Range getSelection(Editor editor) {
    SelectionModel selectionModel = editor.getSelectionModel();
    VisualPosition selectionStartPosition = selectionModel.getSelectionStartPosition();
    VisualPosition selectionEndPosition = selectionModel.getSelectionEndPosition();
    if (selectionStartPosition != null && selectionEndPosition != null) {
      return new Range()
          .setStart(
              new Position()
                  .setLine(selectionStartPosition.line)
                  .setCharacter(selectionStartPosition.column))
          .setEnd(
              new Position()
                  .setLine(selectionEndPosition.line)
                  .setCharacter(selectionEndPosition.column));
    }
    List<Caret> carets = editor.getCaretModel().getAllCarets();
    if (!carets.isEmpty()) {
      Caret caret = carets.get(0);
      Position position =
          new Position()
              .setLine(caret.getLogicalPosition().line)
              .setCharacter(caret.getLogicalPosition().column);
      // A single-offset caret is a selection where end == start.
      return new Range().setStart(position).setEnd(position);
    }
    return null;
  }

  // Sends a textDocument/didChange notification to the agent server.
  public static void informAgentAboutEditorChange(
      @Nullable Editor editor, boolean skipCodebaseOnFileOpened) {
    if (editor == null) {
      return;
    }
    if (editor.getProject() == null) {
      return;
    }
    if (!CodyAgent.isConnected(editor.getProject())) {
      return;
    }
    CodyAgentClient client = CodyAgent.getClient(editor.getProject());
    if (client.server == null) {
      return;
    }
    VirtualFile file = FileDocumentManager.getInstance().getFile(editor.getDocument());
    if (file == null) {
      return;
    }
    TextDocument document =
        new TextDocument()
            .setFilePath(file.getPath())
            .setContent(editor.getDocument().getText())
            .setSelection(getSelection(editor));
    client.server.textDocumentDidChange(document);

    if (client.codebase == null || skipCodebaseOnFileOpened) {
      return;
    }
    client.codebase.onFileOpened(editor.getProject(), file);
  }
}

File: src/main/java/com/sourcegraph/cody/autocomplete/CodyEditorFactoryListener.kt
package com.sourcegraph.cody.autocomplete

import com.intellij.openapi.command.CommandProcessor
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.event.*
import com.intellij.openapi.editor.ex.util.EditorUtil
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.TextEditor
import com.intellij.openapi.fileEditor.impl.FileEditorManagerImpl
import com.intellij.openapi.util.Disposer
import com.sourcegraph.cody.agent.CodyAgent.Companion.getClient
import com.sourcegraph.cody.agent.CodyAgent.Companion.isConnected
import com.sourcegraph.cody.agent.protocol.Position
import com.sourcegraph.cody.agent.protocol.Range
import com.sourcegraph.cody.agent.protocol.TextDocument
import com.sourcegraph.cody.autocomplete.CodyAutocompleteManager.Companion.instance
import com.sourcegraph.cody.vscode.InlineAutocompleteTriggerKind
import com.sourcegraph.cody.vscode.InlineCompletionTriggerKind
import com.sourcegraph.config.ConfigUtil.isCodyEnabled
import com.sourcegraph.utils.CodyEditorUtil.VIM_EXIT_INSERT_MODE_ACTION
import com.sourcegraph.utils.CodyEditorUtil.isEditorValidForAutocomplete
import com.sourcegraph.utils.CodyEditorUtil.isImplicitAutocompleteEnabledForEditor

/**
 * Determines when to trigger completions and when to clear completions.
 *
 * IntelliJ doesn't have a built-in API to register "inline completion providers" similar to VS
 * Code. Instead, we manually listen to editor events like the caret position, selection changes,
 * and document edits.
 */
class CodyEditorFactoryListener : EditorFactoryListener {
  private var selectionListener = CodySelectionListener()
  private var caretListener: CaretListener = CodyCaretListener()

  override fun editorCreated(event: EditorFactoryEvent) {
    if (!isCodyEnabled()) {
      return
    }
    val editor = event.editor
    informAgentAboutEditorChange(editor, skipCodebaseOnFileOpened=false)
    val project = editor.project
    if (project == null || project.isDisposed) {
      return
    }
    val disposable = Disposer.newDisposable("CodyEditorFactoryListener")
    EditorUtil.disposeWithEditor(editor, disposable)
    editor.caretModel.addCaretListener(caretListener, disposable)
    editor.selectionModel.addSelectionListener(selectionListener, disposable)
    editor.document.addDocumentListener(CodyDocumentListener(editor), disposable)
  }

  private class CodyCaretListener : CaretListener {
    override fun caretPositionChanged(e: CaretEvent) {
      if (!isCodyEnabled()) {
        return
      }
      val commandName = CommandProcessor.getInstance().currentCommandName
      if (commandName == VIM_EXIT_INSERT_MODE_ACTION) {
        return
      }
      informAgentAboutEditorChange(e.editor)
      val suggestions = instance
      val editor = e.editor
      if (isEditorValidForAutocomplete(editor) && isSelectedEditor(editor)) {
        suggestions.clearAutocompleteSuggestions(e.editor)
        if (isImplicitAutocompleteEnabledForEditor(editor))
            suggestions.triggerAutocomplete(
                e.editor, e.editor.caretModel.offset, InlineCompletionTriggerKind.AUTOMATIC)
      }
    }
  }

  private class CodySelectionListener : SelectionListener {
    override fun selectionChanged(e: SelectionEvent) {
      if (!isCodyEnabled()) {
        return
      }
      val editor = e.editor
      informAgentAboutEditorChange(editor)
      if (isEditorValidForAutocomplete(editor) && isCodyEnabled() && isSelectedEditor(editor))
          instance.clearAutocompleteSuggestions(editor)
    }
  }

  private class CodyDocumentListener(private val editor: Editor) : BulkAwareDocumentListener {
    override fun documentChangedNonBulk(event: DocumentEvent) {
      if (!isSelectedEditor(editor)) {
        return
      }
      val completions = instance
      completions.clearAutocompleteSuggestions(editor)
      if (isImplicitAutocompleteEnabledForEditor(editor) &&
          isEditorValidForAutocomplete(editor) &&
          !CommandProcessor.getInstance().isUndoTransparentActionInProgress) {
        informAgentAboutEditorChange(editor)
        val changeOffset = event.offset + event.newLength
        if (editor.caretModel.offset == changeOffset) {
          val requestType =
              if (event.oldLength != event.newLength) InlineAutocompleteTriggerKind.Invoke
              else InlineAutocompleteTriggerKind.Automatic
          completions.triggerAutocomplete(
              editor, changeOffset, InlineCompletionTriggerKind.AUTOMATIC)
        }
      }
    }
  }

  companion object {
    /**
     * Returns true if this editor is currently open and focused by the user. Returns true if this
     * editor is in a separate tab or not focused/selected by the user.
     */
    private fun isSelectedEditor(editor: Editor?): Boolean {
      if (editor == null) {
        return false
      }
      val project = editor.project
      if (project == null || project.isDisposed) {
        return false
      }
      val editorManager = FileEditorManager.getInstance(project) ?: return false
      if (editorManager is FileEditorManagerImpl) {
        val current = editorManager.getSelectedTextEditor(true)
        return current != null && current == editor
      }
      val current = editorManager.getSelectedEditor()
      return current is TextEditor && editor == current.editor
    }

    private fun getSelection(editor: Editor): Range? {
      val selectionModel = editor.selectionModel
      val selectionStartPosition = selectionModel.selectionStartPosition
      val selectionEndPosition = selectionModel.selectionEndPosition
      if (selectionStartPosition != null && selectionEndPosition != null) {
        return Range()
            .setStart(
                Position()
                    .setLine(selectionStartPosition.line)
                    .setCharacter(selectionStartPosition.column))
            .setEnd(
                Position()
                    .setLine(selectionEndPosition.line)
                    .setCharacter(selectionEndPosition.column))
      }
      val carets = editor.caretModel.allCarets
      if (!carets.isEmpty()) {
        val caret = carets[0]
        val position =
            Position()
                .setLine(caret.logicalPosition.line)
                .setCharacter(caret.logicalPosition.column)
        // A single-offset caret is a selection where end == start.
        return Range().setStart(position).setEnd(position)
      }
      return null
    }

    // Sends a textDocument/didChange notification to the agent server.
    fun informAgentAboutEditorChange(editor: Editor?, skipCodebaseOnFileOpened: Boolean = true) {
      if (editor == null) {
        return
      }
      if (editor.project == null) {
        return
      }
      if (!isConnected(editor.project!!)) {
        return
      }
      val client = getClient(editor.project!!)
      if (client.server == null) {
        return
      }
      val file = FileDocumentManager.getInstance().getFile(editor.document) ?: return
      val document =
          TextDocument()
              .setFilePath(file.path)
              .setContent(editor.document.text)
              .setSelection(getSelection(editor))
      client.server!!.textDocumentDidChange(document)
      if (client.codebase == null || skipCodebaseOnFileOpened) {
        return
      }
      client.codebase!!.onFileOpened(editor.project!!, file)
    }
  }
}
